### ISSUE TITLE
`crt-ignore-err all` broken after v2.3.9

### ISSUE THREAD

**Issue by suvl** [haproxy/haproxy#1425]
### Detailed Description of the Problem

We're having issues upgrading from v2.3.9 onwards (all new versions are the same) because the behaviour of `crt-ignore-err all` appears to have changed.

We have this `bind` line:

```
bind abns@haproxy-clientcert accept-proxy ssl crt /etc/ssl/certs/company/ ca-file /etc/ssl/certs/companyca/all-ca-certs.pem verify required crt-ignore-err all alpn h2,http/1.1 curves X25519:P-256:secp384r1
```

Now for informational purposes, the `all-ca-certs.pem` has:
* 7 CAs from different providers and different configurations
* all of them are still valid (the soonest one will expire will be 2031, the latest 2051)
* some were created long ago (2011) others very recently (2018)
* two are `sha1WithRSAEncryption` signed, the other five are `sha256WithRSAEncryption`

We know for a fact that there are a lot of clients with expired certificates connecting to us, and we cannot replace those certificates without replacing expensive equipment. In an effort to keep validating most clients while allowing these expired ones, we're having some ACLs to try and keep some security:

```
    acl is_problematic_client ssl_c_i_dn -m str "/O=Problematic OU/CN=Problematic CA"
    acl is_ssl_err ssl_c_verify gt 0
    http-request deny if is_ssl_err !is_problematic_client 
```

Up until version 2.3.9 this worked quite well but since we updated to 2.3.10 onwards (and all 2.4 versions) this behaviour broke and every time the connection is denied, haproxy logs:
```
Oct 22 16:14:43 localhost haproxy[9]: 101.95.0.25:50034 [22/Oct/2021:16:14:43.156] fe-ssl-clientcert/1: SSL client certificate not trusted 
```

### Expected Behavior

I can see no changelog about the behaviour of `crt-ignore-err all` so we are expecting haproxy to let through the request, as it does in v2.0.0 to v2.3.9.

### Steps to Reproduce the Behavior

Not really sure but:
1. have two CAs, one signed with `sha1WithRSAEncryption`, the other `sha256WithRSAEncryption` (not sure if this is relevant or one CA would suffice)
2. have clients with expired client certificates (the `ssl_c_verify` value for these is `21`)
3. Try and get these requests through without logging "SSL client certificate not trusted"

### Do you have any idea what may have caused this?

some change in the openssl api calls are not taking into account `crt-ignore-err all`

### What is your configuration?

```haproxy
global
    log         127.0.0.1 local2
    chroot      /var/lib/haproxy
    pidfile     /var/run/haproxy.pid
    maxconn     199000
    daemon
    nbthread 4
    master-worker
    stats socket /var/lib/haproxy/stats
    stats socket *:1999 level admin
    stats socket /var/run/haproxy.sock mode 600 level admin
    server-state-file /etc/haproxy/haproxy.state

    ssl-default-bind-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384
    ssl-default-bind-options no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets

    ssl-default-server-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384
    ssl-default-server-options no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets

    tune.ssl.default-dh-param 2048
    ssl-dh-param-file /data/files/dhparam.pem
    tune.listener.multi-queue on
    tune.maxpollevents 200
    tune.maxaccept 32
    tune.ssl.cachesize 20000

    ssl-engine rdrand
    ssl-mode-async

defaults
    mode                    http
    log                     global
    option                  httplog
    option                  dontlognull
    option                  log-separate-errors
    option                  http-server-close
    option forwardfor       except 127.0.0.0/8
    option                  redispatch
    option                  tcp-smart-connect
    retries                 3
    timeout http-request    10s
    timeout queue           1m
    timeout connect         10s
    timeout client          1m
    timeout server          1m
    timeout http-keep-alive 10s
    timeout check           10s
    maxconn                 256000
    fullconn                256000
    backlog                 256000

# main listening frontend is via TCP for loadbalancing based on SNI

frontend web_443
    bind *:443
    mode tcp
    maxconn 128000
    tcp-request inspect-delay 1s
    stick-table type ip size 1m expire 10s store  gpc0,conn_rate(10s)
    acl source_is_abuser src_get_gpc0 gt 0
    tcp-request connection reject if source_is_abuser
    acl conn_rate_abuse  sc1_conn_rate gt 350
    acl mark_as_abuser   sc1_inc_gpc0  ge 0
    tcp-request connection track-sc1 src
    tcp-request connection reject if conn_rate_abuse mark_as_abuser
    tcp-request content capture req.ssl_sni len 20
    tcp-request content accept if { req_ssl_hello_type 1 }
    log-format "%ci:%cp [%t] %ft SNI:%[capture.req.hdr(0)] %b/%s %Tw/%Tc/%Tt %B %ts %ac/%fc/%bc/%sc/%rc %sq/%bq {%[ssl_c_verify],%{+Q}[ssl_c_s_dn(cn)],%{+Q}[ssl_c_i_dn]}" 

    # simplifying this a bit just for this issue
    use_backend recir_clientcertenabled if { req.ssl_sni -i example.com }
    use_backend some_other_backend if { req.ssl_sni -i tcprequest.example.com }

backend recir_clientcertenabled
    mode tcp
    server loopback-for-tls abns@haproxy-clientcert send-proxy-v2
```

### Output of `haproxy -vv`

```plain
** Running (working) version **

HA-Proxy version 2.3.9-53945bf 2021/03/30 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2022.
Known bugs: http://www.haproxy.org/bugs/bugs-2.3.9.html
Running on: Linux 3.10.0-1062.12.1.el7.x86_64 #1 SMP Tue Feb 4 23:02:59 UTC 2020 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_LINUX_TPROXY=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_TFO=1
  DEBUG   =

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=4).
Built with OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
Running on OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with the Prometheus exporter as a service
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.34 2019-11-21
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 9.3.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services : prometheus-exporter
Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [COMP] compression
        [TRACE] trace

** Failing Version **

HA-Proxy version 2.3.10-4764f0e 2021/04/23 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2022.
Known bugs: http://www.haproxy.org/bugs/bugs-2.3.10.html
Running on: Linux 3.10.0-1062.12.1.el7.x86_64 #1 SMP Tue Feb 4 23:02:59 UTC 2020 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_LINUX_TPROXY=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_TFO=1
  DEBUG   =

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=4).
Built with OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
Running on OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with the Prometheus exporter as a service
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.34 2019-11-21
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 9.3.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services : prometheus-exporter
Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [COMP] compression
        [TRACE] trace
```

### Additional Information

haproxy running on docker, no kubernetes. docker images' versions:
1. haproxytech/haproxy-ubuntu:2.3.9
2. haproxytech/haproxy-ubuntu:2.3.10

**Comment by suvl** [comment]
Would benefit greatly from #693

**Comment by wlallemand** [comment]
That's strange, there was no bugfix backported regarding this, but there were some regarding the SSL stack, it could be the side effect of another fix. I'll try to reproduce.

**Comment by wlallemand** [comment]
I just check your configuration, and I'm surprised it even worked:

- you don't have any SSL endpoint in your haproxy configuration

**Comment by wlallemand** [comment]
Are you sure it worked correctly? `10` is the correct code for `X509_V_ERR_CERT_HAS_EXPIRED`, the `21` code is for `X509_V_ERR_UNABLE_TO_VERIFY_LEAF_SIGNATURE` which seems to be a chain problem.

**Comment by suvl** [comment]
> Are you sure it worked correctly? `10` is the correct code for `X509_V_ERR_CERT_HAS_EXPIRED`, the `21` code is for `X509_V_ERR_UNABLE_TO_VERIFY_LEAF_SIGNATURE` which seems to be a chain problem.

`crt-ignore-err all` would need to ignore both, right? I'm not sure where the expired cert is, may be in the middle of the chain. Even so, with version 2.3.9 it let's the request through and with any later version it doesn't and that cannot be.

**Comment by EmericBr** [comment]
Just for note: ssl-mode-async is useless on engine rdrand, since this engine doesn't use a device and was not blocking. ssl-mode-async was designed to work using intel quick assist engine.

**Comment by wtarreau** [comment]
Is anything still needed in this issue ?

**Comment by suvl** [comment]
@wtarreau what you mean? If there is a need for more info or if this still happens?

**Comment by wtarreau** [comment]
I'm seeing that the issue has not made any progress in any direction over the last few months, so it's possible that the SSL maintainers have no more ideas and might need more info, but I don't know. I hate leaving pending issues that do not make progress, we should at the very least reach a point such as unreproduceable, fixed, works as designed or future fix. But each participant waiting for the other one looks like a deadlock. I can understand that the forthcoming release keeps everyone busy but I would like this issue to be revisited next week or so, so that we can conclude something.

**Comment by wlallemand** [comment]
> > Are you sure it worked correctly? `10` is the correct code for `X509_V_ERR_CERT_HAS_EXPIRED`, the `21` code is for `X509_V_ERR_UNABLE_TO_VERIFY_LEAF_SIGNATURE` which seems to be a chain problem.
> 
> `crt-ignore-err all` would need to ignore both, right? I'm not sure where the expired cert is, may be in the middle of the chain. Even so, with version 2.3.9 it let's the request through and with any later version it doesn't and that cannot be.

If that's a chain problem that's the job of "ca-ignore-err". But it looks like your chain doesn't work anymore, and we didn't change anything about this in HAProxy.

The 21 error code is X509_V_ERR_UNABLE_TO_VERIFY_LEAF_SIGNATURE, so my guess is that your distribution disabled sha1WithRSAEncryption between your updates.

**Comment by suvl** [comment]
@wlallemand what you mean by "my distribution"? I solely use the official docker images.

**Comment by wlallemand** [comment]
@suvl I meant the "ubuntu" images. Are you able to try with the 2.3.9 and 2.3.10 alpine images ?

**Comment by suvl** [comment]
@wlallemand yep even with the alpine images it still happens.

However, yesterday from your reply I got busy and found this reply from the openssl issues: https://github.com/openssl/openssl/issues/3558#issuecomment-383154880

So I've added the `@SECLEVEL=0` string to the end of our cypher list and I got it working. I was able to upgrade from 2.3.9 to 2.5.7 and still keep all our customer devices connected. No errors from the logs.

I understand now that this was a change on the SSL engine and not haproxy. Nevertheless I want to thank @wlallemand and @wtarreau for keeping up with me 😄 closing now. Good job 👍🏽
